SignalServiceKit/src/Loki/Protocol/Friend Requests/FriendRequestProtocol.swift

```diff
@@ -40,55 +62,126 @@ public final class FriendRequestProtocol : NSObject {
+        // Otherwise don't allow attachments at all
         return false
     }
 
+    @objc(getFriendRequestUIStateForThread:)
+    public static func getFriendRequestUIState(for thread: TSThread) -> FriendRequestUIState {
+        // Friend requests have nothing to do with groups
+        guard let thread = thread as? TSContactThread else { return .none }
+        // If this is a note to self then we don't want to show the friend request
+        if thread.isNoteToSelf() { return .none }
+        var friendRequestStatuses: [LKFriendRequestStatus] = []
+        storage.dbReadConnection.read { transaction in
+            let linkedDevices = LokiDatabaseUtilities.getLinkedDeviceHexEncodedPublicKeys(for: thread.contactIdentifier(), in: transaction)
+            friendRequestStatuses = linkedDevices.map { device in
+                return storage.getFriendRequestStatus(for: device, transaction: transaction)
+            }
+        }
+
+        if friendRequestStatuses.contains(where: { $0 == .friends }) { return .friends }
+        if friendRequestStatuses.contains(where: { $0 == .requestReceived }) { return .received }
+        if friendRequestStatuses.contains(where: { $0 == .requestSent }) { return .sent }
+
+        return .none
+    }
+
     // MARK: - Sending
-    @objc(acceptFriendRequestFrom:in:using:)
-    public static func acceptFriendRequest(from hexEncodedPublicKey: String, in thread: TSThread, using transaction: YapDatabaseReadWriteTransaction) {
+    @objc(acceptFriendRequestFromHexEncodedPublicKey:using:)
+    public static func acceptFriendRequest(from hexEncodedPublicKey: String, using transaction: YapDatabaseReadWriteTransaction) {
         // Accept all outstanding friend requests associated with this user and try to establish sessions with the
         // subset of their devices that haven't sent a friend request.
-        let linkedDeviceThreads = LokiDatabaseUtilities.getLinkedDeviceThreads(for: hexEncodedPublicKey, in: transaction) // This doesn't create new threads if they don't exist yet
-        // FIXME: Capture send failures
-        for thread in linkedDeviceThreads {
-            if thread.hasPendingFriendRequest {
-                sendFriendRequestAcceptanceMessage(to: thread.contactIdentifier(), in: thread, using: transaction) // NOT hexEncodedPublicKey
-                thread.saveFriendRequestStatus(.friends, with: transaction)
-            } else {
-                MultiDeviceProtocol.getAutoGeneratedMultiDeviceFRMessageSend(for: thread.contactIdentifier(), in: transaction) // NOT hexEncodedPublicKey
-                .done(on: OWSDispatch.sendingQueue()) { autoGeneratedFRMessageSend in
-                    let messageSender = SSKEnvironment.shared.messageSender
-                    messageSender.sendMessage(autoGeneratedFRMessageSend)
+        guard ECKeyPair.isValidHexEncodedPublicKey(candidate: hexEncodedPublicKey) else {
+            assertionFailure("Invalid session ID \(hexEncodedPublicKey)")
```

Comment:
Don't know if we need to crash on this

Reply:
We should make it crash when debugging. These assertions should be stripped out in production so it shouldn't make a difference there.

Reply:
Okay, yeah. This would've made it crash in production as well though :)